spec/orders.mediawiki

```diff
@@ -720,7 +724,9 @@ When a client has redeemed their contract, they will notify the server.
 |-
 | matchid    || string    || the match ID
 |-
-| coinid    || string || hex-encoded coin ID
+| coinid     || string || hex-encoded coin ID
+|-
+| secret     || string || the hex-encoded swap contract secret
```

Comment:
Since the redemption will not be a wallet transaction for the counter-party, we shouldn't count on their ability to access the transaction data. The `ExchangeWallet` of course provides the `FindRedemption` method, so the client's funds are not at risk if the counter-party fails to provide a valid secret, but `FindRedemption` is quite costly and should be avoided when possible.

Reply:
You're referring to finding the actual redemption transaction, right?  But what's the motivation for sending the secret?  The taker can extract the secret from the maker's redemption with `FindKeyPush`.  If the secret becomes part of the signed ack messages, that's more to store in the server's DB, and it's also more work for the server to verify the secret in the message matches the secret in the redemption.  Related comment in `(*Swapper).processRedeem`:

https://github.com/decred/dcrdex/blob/4a26dc11871ddff16ea5dcb83fc1c384190cfb7f/server/swap/swap.go#L944-L950

Reply:
> The taker can extract the secret from the maker's redemption with `FindKeyPush`

The taker needs the maker's redemption transaction in order to do that, but since none of the outputs in the maker's redemption transaction go to the taker, the taker's wallet will not know about it, and would have to scan the blockchain to find it. Technically the taker does have a revocation address in the previous outpoint's pubkey script being spent in the revocation, but I don't think that's enough for the wallet to index the transaction. 

At least I think that's how it works.

Reply:
> the taker's wallet will not know about it, and would have to scan the blockchain to find it.

Oh, I think you're referring to the limitation of `gettransaction`, the wallet command.  `getrawtransaction` is a node command that will find *any* txn, and it's passed through the wallet to the node, so no separate node connection is needed to use it.  The counterparties should be able to get all the txn info without the server's help, and I think `getrawtransaction` will do it.  I'm not sure how this will work with lite clients, but there would be a similar solution with compact filters.

Reply:
> If the node has the txindex enabled.

Yeah. I meant to mention that part. `txindex` cannot be used with `prune` either, so the only option to use the DEX right now is storing the entire BTC blockchain with `txindex` enabled. This change is looking forward to dropping that restriction for the client.

Reply:
> without the server's help

The secret originates with the counter-party, not the server. They send it with their `redeem` request, and the server passes it along in the `redemption` request. The server might want to validate it though.

Reply:
But even if the secret is sent to the taker, the taker still needs to confirm that it is correct.  The idea is that they attempt to use the secret in their redemption as a way to confirm it I guess?

My main hesitation to send the secret is that it puts a pretty significant burden on the server to validate that secret, whereas previously it was all the client's responsibility.  Do you think the server can skip secret validation (as alluded to in the linked comment)?  Is there a simpler way to do it other than checking the redemptions inputs to identify the one that spends the contract output, and then extracting the secret from the sig script?

Reply:
> The secret originates with the counter-party, not the server. They send it with their `redeem` request, and the server passes it along in the `redemption` request. The server might want to validate it though.

I just mean that all a counter-party needs is the coinID and then they can pull the tx and extract the secret.  The downsides of the server relaying the secret are: (1) a little extra bandwidth, (2) the need to store it in the DB also since we are now storing all the data corresponding to the sig acks, and (3) the server probably would want to validate the secret before relaying it.

Reply:
> The idea is that they attempt to use the secret in their redemption as a way to confirm it I guess?

In my unpublished match negotiation branch, I've added a `ValidateSecret(secret, secretHash []byte)` method to `ExchangeWallet`. The secret hash will have already been pulled when the `audit` request was received. Lot easier when it's a utxo.

> Is there a simpler way to do it other than checking the redemptions inputs to identify the one that spends the contract output, and then extracting the secret from the sig script?

The server already has a validated contract in hand from the `init` request, so no RPC calls are needed. My solution has the `asset.Backend` getting its own `ValidateSecret(secret, contract []byte)` method. The `Contract` type from #202 is already validating the script and the `Contract` type will provide a getter for convenience.

> the need to store it in the DB also since we are now storing all the data corresponding to the sig acks

I guess that's true. Damn.

Reply:
That makes sense.  Thanks for the preview of the negotiation work.